src/persist-client/src/bin/persist_open_loop_benchmark.rs

```diff
@@ -0,0 +1,449 @@
+                            None => {
+                                let records_sent = (batches_sent as usize) * args.batch_size;
+                                let finished = format!(
+                                    "Data generator {} finished after {} ms and sent {} records",
+                                    idx, elapsed_millis, records_sent
+                                );
+                                return Ok(finished);
+                            }
+                        };
+                        trace!("data generator {} wrote a batch", idx);
+                        // TODO: clearly specify exact error type. This can only error
+                        // if the receiver has been dropped.
+                        if let Err(_) = tx.send(batch) {
+                            bail!("receiver unexpectedly dropped");
+                        }
+                        batches_sent += 1;
+                    }
+
+                    if elapsed - prev_log > FIVE_SECONDS {
+                        let records_sent = (batches_sent as usize) * args.batch_size;
+                        info!(
+                            "After {} ms data generator {} has sent {} records.",
+                            elapsed_millis, idx, records_sent
+                        );
+                        prev_log = elapsed;
+                    }
+                }
+            },
+        );
+
+        handles.push(data_generator_handle);
+        let b = Arc::clone(&barrier);
+
+        let writer_handle = mz_ore::task::spawn(|| format!("writer-{}", idx), async move {
+            trace!("writer {} waiting for barrier", idx);
+            b.wait().await;
+            info!("starting writer {}", idx);
+
+            // Max observed latency for BenchmarkWriter::write.
+            let mut max_write_latency = Duration::from_millis(0);
+
+            // The last time we emitted progress information to stdout, expressed
+            // as a relative duration from start.
+            let mut prev_log = Duration::from_millis(0);
+            let mut batches_written = 0;
+
+            loop {
+                let batch = match rx.recv().await {
+                    Some(batch) => batch,
+                    None => {
+                        let elapsed = start.elapsed();
+                        let records_written = batches_written * args.batch_size;
+                        let finished = format!("Writer {} finished after {} ms and wrote {} records. Max write latency {} ms.",
+                          idx, elapsed.as_millis(), records_written, max_write_latency.as_millis());
+                        return Ok(finished);
+                    }
+                };
+
+                trace!("writer {} received a batch. writing", idx);
+                let write_start = Instant::now();
+                writer.write(batch).await?;
+                batches_written += 1;
+                let write_latency = write_start.elapsed();
```

Comment:
This only gives us the latency of writing each batch. I think the more useful measure is how far we're lagging behind on the queue of batches that we need to write.

Reply:
so we're reporting the end to end lag between "how much we expect to have read" vs "how much we actually read" which is I think the most useful metric here. I agree that a writer lag metric would be useful -- for now I'm kind of inclined to say that the writer lag is derivable from the information the writer logs out ("after <foo> ms the writer wrote <bar> records") and we can revisit if we want to track writer lag explicitly.

Reply:
crucially - we're measuring the end-to-end lag based on the expected data generation schedule and not the schedule of which records have been written so far.

Reply:
A scratch that, we record how much we lag in terms of number of records.

Reply:
🙈  I responded while I didn't see your comments yet.